### ISSUE TITLE
Reproducing Results

### ISSUE THREAD

**Issue by nilsleh** [aangelopoulos/im2im-uq#5]
Hi @aangelopoulos and @apsk14  , 

Thanks for your great work, I am trying to reproduce the paper results, because I would love to add the img2img conformal method to the [Lightning-UQ-Box](https://github.com/lightning-uq-box/lightning-uq-box).

I have run your code as instructed in the README.md via the sweep for the FastMRI Dataset (singlecoil_train), but the data normalization is doing some strange things. The normalization parameters being computed (taken from the cached `norm_params.pickle` file) are as follows:

```python
{
    'input_max': 30.512170791625977, 
    'input_min': -0.604563295841217, 
    'input_mean': 0.06673449277877808, 
    'input_std': 0.8941464991811945, 
    'output_max': 7.970857620239258, 
    'output_min': -0.0070489561185240746, 
    'output_mean': 0.1550026535987854, 
    'output_std': 0.2108842432498932
}
```

However, these values are quiet large given the data and have the effect that the inputs and targets have some strange statistics after being normalized (both with `standard` and `min_max`) when I check them during the loss or eval computation

```
GET THE METRICS INCLUDING SPATIAL MISCOVERAGE
dataset is map-style (not resettable)
Validation output 0
mean: -0.07451456785202026
std: 9.658681665314361e-05
min: -0.07463472336530685
max: -0.07418996095657349
model targets
mean: 0.0008978385012596846
std: 1.147075727203628e-05
min: 0.0008843542891554534
max: 0.0009457280393689871

Validation output 1
mean: -0.0745149627327919
std: 9.722328832140192e-05
min: -0.07463472336530685
max: -0.07417760789394379
model targets
mean: 0.0008978385012596846
std: 1.147075727203628e-05
min: 0.0008843542891554534
max: 0.0009457280393689871

Validation output 2
mean: -0.07451476156711578
std: 9.685881377663463e-05
min: -0.07463475316762924
max: -0.07417760789394379
model targets
mean: 0.0008978385012596846
std: 1.147075727203628e-05
min: 0.0008843542891554534
max: 0.0009457280393689871
```

Essentially, the inputs and output get normalized to basically constant values, with no variation in pixel values.

I went down this path because, I was a bit astound by the very low mse value and interval size reported in the paper, but also in the metrics that are being computed:

```
# IMG2IMG Results:
# Risk: 0.0847138
# Mean size: 0.001457
# Spearman: 0.231718
# Strat Risk: [0.086, 0.0839, 0.0839, 0.0851]
# MSE: 7.1185ee-06
# Spatial Miscoverage: mu 0.08471, sigma: 0.271718, min: 0.0, max: 1.0
``` 

I have a script with which I compute normalization statistics for datasets and get statistics that are quiet different and orders of magnitude lower:

```python
{
    "input_min": 1.325229959547869e-06,
    "input_max": 0.0003247384447604418,
    "input_mean": 4.692213955797128e-05,
    "input_std": 7.761315646348521e-05,
    "output_min": 2.350711156395846e-06,
    "output_max": 0.00033284907112829387,
    "output_mean": 4.8656625481538975e-05,
    "output_std": 7.794056000420824e-05
}
 ```
However, when using those, the image statistics after normalization seem more reasonable (more pixel variance):

```
Validation output 0
mean: -0.4299814701080322
std: 0.14215296506881714
min: -0.6043229699134827
max: 0.26245826482772827
model targets
mean: 0.04337747395038605
std: 0.03413794934749603
min: -0.00697715999558568
max: 0.32631415128707886

Validation output 1
mean: -0.4297109842300415
std: 0.14215296506881714
min: -0.6041231155395508
max: 0.2342386245727539
model targets
mean: 0.04337747395038605
std: 0.03413794934749603
min: -0.00697715999558568
max: 0.32631415128707886

Validation output 2
mean: -0.4299924373626709
std: 0.14188271760940552
min: -0.6043405532836914
max: 0.2342386245727539
model targets
mean: 0.04337747395038605
std: 0.03413794934749603
min: -0.00697715999558568
max: 0.32631415128707886
```

I haven't found the exact issue for why I am observing the normalized images I am seeing with your script (although I am suspecting it in the `RunningStats` object` at the moment), but was hoping that I might have missed something obvious, before diving down deeper into the rabbit hole. I was hoping that you might be able to confirm my observations on your end or have any other pointers. Thanks in advance :)

**Comment by aangelopoulos** [comment]
Hey @nilsleh !! I'm so happy to hear that you're interested in the pixelwise UQ stuff! It's such a fun problem to work on :D 

Regarding the normalization, it's very possible we have an error. It does look like the standard deviation would ideally be closer to 1. Although in your case, there also looks like there's something weird going on, because the mean is always around -0.429... shouldn't it be close to 0?

I remember that there were some issues with the FastMRI dataset, but I don't recall exactly what they were. @apsk14 wrote these preprocessing steps, right? Amit, do you know what's going on and why these values are tiny?

**Comment by nilsleh** [comment]
@aangelopoulos Thanks for your fast reply. 

I am afraid I might have overlooked something I did, because I cloned the repo again to a new cleaned environment and rerunning the script actually yields normalization statistics that have the same order of magnitude that I computed, and hence I also don't see the constant values with no pixel variance anymore. I should have done that before writing the issue, apologies for that. Not sure what happened there.

**Comment by aangelopoulos** [comment]
No problem at all. Glad there's no issue... phew! 😅

Thanks again for reaching out! This is super important stuff, and it's all hands on deck 🙌🏼

**Comment by nilsleh** [comment]
Sorry for commenting on the closed issue, but thought I would let you know that I was able to reproduce your results with the implementation that I just merged into main of the [Lightning-UQ-Box](https://github.com/lightning-uq-box/lightning-uq-box), will be part of the `0.2.0` release coming soon. Used the following script:

```python
import os
from datetime import datetime

from lightning import Trainer
from lightning.pytorch.loggers import CSVLogger
from lightning_uq_box.models.img2img_conformal import QuantileLoss
from lightning_uq_box.uq_methods import Img2ImgConformal, QuantilePxRegression

from core.models.add_uncertainty import add_uncertainty
from core.models.trunks.unet import UNet
from datamodule import FastMRIDatamodule

# datamodule
path = "/mnt/SSD2/nils/projects/mri/singlecoil_train"
dm = FastMRIDatamodule(path, batch_size=64, num_workers=8)
dm.setup("fit")

# set up Unet Model from the paper
model = UNet(n_channels_in=1, n_channels_out=1)

model_params = {
    "uncertainty_type": "quantiles",
    "q_lo": 0.05,
    "q_hi": 0.95,
    "q_lo_weight": 1.0,
    "q_hi_weight": 1.0,
    "mse_weight": 1.0,
}
model = add_uncertainty(model, model_params)

# fit the quantile regression model
quantile_model = QuantilePxRegression(
    model=model,
    loss_fn=QuantileLoss,
)

# setup experiment directory to save results
timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
exp_name = "reproduce"
my_dir = os.path.join("my_experiments", f"{timestamp}_{exp_name}")
if not os.path.exists(my_dir):
    os.makedirs(my_dir)

loggers = [
    CSVLogger(my_dir, name="csv_logs"),
]

trainer = Trainer(
    max_epochs=1,
    accelerator="gpu",
    devices=[0],
    default_root_dir=my_dir,
    logger=loggers,
)

# train the quantile regression model
trainer.fit(quantile_model, datamodule=dm)

# setup conformal model
conformal_model = Img2ImgConformal(quantile_model, alpha=0.1)

calib_loader = dm.calib_dataloader()

# conformal calibration step
trainer.validate(conformal_model, dataloaders=calib_loader)

# test with conformal model
trainer.test(conformal_model, datamodule=dm)
```

If you have any feedback or suggestions, feel free to reach out. Thanks again for the open code and being so helpful (and sorry for scaring you with this issue). Looking to transcribe Conformal Risk Control next!

**Comment by aangelopoulos** [comment]
This is incredible! I am looking forward to using your library @nilsleh  :)

It's going to be very useful to have a reference implementation for this and other methods that's compatible with torch. I think you're doing a great community service. Congrats on your awesome work.
